src/bio/bio_wal.h

```diff
@@ -37,41 +37,38 @@ struct wal_header {
 	uint64_t	wh_tot_blks;	/* WAL blob capacity, in blocks */
 	uint64_t	wh_ckp_id;	/* Last check-pointed transaction ID */
 	uint64_t	wh_commit_id;	/* Last committed transaction ID */
-	uint64_t	wh_unused_id;	/* Next unused transaction ID */
+	uint32_t	wh_ckp_blks;	/* blocks used by last check-pointed transaction */
+	uint32_t	wh_commit_blks;	/* blocks used by last committed transaction */
 	uint64_t	wh_padding2;	/* Reserved */
 	uint32_t	wh_padding3;	/* Reserved */
 	uint32_t	wh_csum;	/* Checksum of this header */
 };
 
-enum wal_hdr_flags {
-	WAL_HDR_FL_TAIL	= (1UL << 0),	/* The tail entry is in current block */
-};
+/*
+ * WAL transaction starts with a header (wal_trans_head), and there will one or multiple
+ * entries (wal_trans_entry) being placed immediately after the header, the payload data
+ * from entries are concatenated after the last entry, the tail (wal_trans_tail) will be
+ * placed after the payload (or the last entry when there is no payload).
+ *
+ * When the transaction spans multiple WAL blocks, the header will be duplicated to the
+ * head of each block.
+ */
 
 /* WAL transaction header */
 struct wal_trans_head {
 	uint32_t	th_magic;
 	uint32_t	th_gen;		/* WAL re-format timestamp */
-	uint32_t	th_len;		/* Total length of transaction entries in WAL */
-	uint32_t	th_flags;	/* See wal_hdr_flags */
 	uint64_t	th_id;		/* Transaction ID */
-};
-
-/* XXX to be adjusted according to umem action types */
-enum wal_op_type {
-	WAL_OP_MEMCPY	= 0,		/* memcpy data to meta blob */
-	WAL_OP_MEMMOVE,			/* memmove data on meta blob */
-	WAL_OP_MEMSET,			/* memset data on meta blob */
-	WAL_OP_DATA_CSUM,		/* checksum of data on data blob */
-	WAL_OP_MAX,
+	uint32_t	th_tot_ents;	/* Total entries */
+	uint32_t	th_tot_payload;	/* Total payload size in bytes */
 };
 
 /* WAL transaction entry */
 struct wal_trans_entry {
 	uint64_t	te_off;		/* Offset within meta blob, in bytes */
-	uint32_t	te_type;	/* Operation type, see wal_op_type */
-	uint32_t	te_len;		/* Data length in bytes */
-	/* Copied data for OP_MEMCPY, Offset for OP_MEMMOVE, Checksum for DATA_CSUM */
-	uint8_t		te_data[0];
+	uint16_t	te_type;	/* Operation type, see UMEM_ACT_XXX */
+	uint16_t	te_len;		/* Data length in bytes */
```

Comment:
so each data length will not exceed 4K? just wondering if this could be extended eg we want to support 64K page size.

Reply:
right, the data length for each action can't exceed 64k. This length is the meta data being changed by single action, it's irrelevant to page size.